File: TooManyReviewers/SimpleCache.swift
//
//  SimpleCache.swift
//  TooManyReviewers
//
//  Created by MeowCat on 2025/2/3.
//

class SimpleCache<K: Hashable, V> {
    private var cache = [K: V]()
    
    private let fn: (K) -> V
    
    init(fn: @escaping (K) -> V) {
        self.fn = fn
    }
    
    subscript(_ key: K) -> V {
        if let cached = cache[key] {
            return cached
        }
        let value = fn(key)
        cache[key] = value
        return value
    }
}

File: src/simple_cache.ts
export class SimpleCache<K, V> {
  private cache: Map<K, V> = new Map();
  
  constructor(private fn: (key: K) => V) { }
  
  get(key: K): V {
    let cached = this.cache.get(key);
    if (cached !== undefined) { return cached; }
    let value = this.fn(key);
    this.cache.set(key, value);
    return value;
  }
}
